### PR TITLE
[Add] Support VIS evaluation for YouTube-VIS task.

### PR DESCRIPTION
**Pull request by pixeli99**
There are three main modifications here.
1. Add `mmtrack/core/evaluation/eval_vis.py`
2. When converting the dataset to coco format, add `width` and `height` info to the key of `video`.
3. Override the function of `evaluate` in `mmtrack/datasets/youtube_vis_dataset.py`.

For the first part, the JSON format used for training is converted into the native JSON format of YouTube-VIS for calculation.

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
The overall structure needs to be adjusted, we should try to reuse the current code and only add some necessary functions. Some suggestions are as the following:
1. Try to only add a file containing the `YTVOSeval` class, because we only want to add the function of evaluation, rather than dataset loading and paring functions.
2. To my observation, the function of `YTVOS` is similar to the `CocoVideoDataset` and 'YoutubeVISDataset'. Can we only add some necessary functions into the 
`YoutubeVISDataset`  and overload some functions of `YoutubeVISDataset`?
3.  Is the function `get_vis_json` necessary?  To my observation, its contents of `annotations`  are similar to the `tools/convert_datasets/youtubvis/youtubevis2coco.py`
4. Some naming problems. We try not to use uppercase letters for the name of normal variables.

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 1]
```diff
@@ -137,3 +145,88 @@ def format_results(self,
...
+        frame_id = defaultdict(list)
+        len_video = defaultdict(list)
+        for ann_info in ori_anns['annotations']:
+            instance_info[ann_info['instance_id']].append(ann_info)
+
+        for img_info in ori_anns['images']:
+            frame_id[img_info['id']] = img_info['frame_id']
+
+            len_video[img_info['video_id']] = max(1, img_info['frame_id'] + 1)
+        for idx in instance_info:
```
Comment (JingweiZhang12) [inline mmtrack/datasets/youtube_vis_dataset.py:212]:
```suggestion
        for ins_id in instances_infos:
            instance_infos =  instance_infos[ind_id]
            cur_video_len = len_videos[instance_infos[0]['video_id']]
        
        
```

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 2]
```diff
@@ -137,3 +145,88 @@ def format_results(self,
...
+        """
+
+        VIS = defaultdict(list)
+        ori_anns = mmcv.load(self.ann_file)
+        VIS['categories'] = copy.deepcopy(ori_anns['categories'])
+        VIS['videos'] = copy.deepcopy(ori_anns['videos'])
+
+        instance_info = defaultdict(list)
+        frame_id = defaultdict(list)
+        len_video = defaultdict(list)
```
Comment (JingweiZhang12) [inline mmtrack/datasets/youtube_vis_dataset.py:204]: len_videos

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 3]
```diff
@@ -137,3 +145,88 @@ def format_results(self,
...
+                ``bboxes``, ``video_id``, ``areas``, ``id`` and ``iscrowd``.
+        """
+
+        VIS = defaultdict(list)
+        ori_anns = mmcv.load(self.ann_file)
+        VIS['categories'] = copy.deepcopy(ori_anns['categories'])
+        VIS['videos'] = copy.deepcopy(ori_anns['videos'])
+
+        instance_info = defaultdict(list)
+        frame_id = defaultdict(list)
```
Comment (JingweiZhang12) [inline mmtrack/datasets/youtube_vis_dataset.py:203]:
```suggestion
        frame_id_mapping = defaultdict(int) # mapping from image_id to frame_id
```

Comment (lpxgis) [reply mmtrack/datasets/youtube_vis_dataset.py:203]: Very good suggestion, thank you😀

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 4]
```diff
@@ -172,61 +172,64 @@ def evaluate(self, results, metric=['track_segm'], logger=None):
...
 
         VIS = defaultdict(list)
         ori_anns = mmcv.load(self.ann_file)
         VIS['categories'] = copy.deepcopy(ori_anns['categories'])
         VIS['videos'] = copy.deepcopy(ori_anns['videos'])
 
-        instance_info = defaultdict(list)
-        frame_id = defaultdict(list)
-        len_video = defaultdict(list)
+        instance_infos = defaultdict(list)
```
Comment (JingweiZhang12) [inline mmtrack/datasets/youtube_vis_dataset.py:204]: Deprecate the old code,  we could use the coco api to realize this function as the following:
```python
VIS = defaultdict(list)
the 'categories' and 'videos' has already loaded in this dataset, don't need to reload it.
for ins_id, img_ids in self.instancesToImgs.items():
    instance_infos = []
    for img_id in img_ids:
           for ann in self.coco.imgToAnns[img_id]:
                if ann['instance_id'] == instance_id:
                    add some code to load ann information of this instance
    VIS['annotations'].append(instance_infos)
```

**Comment by JingweiZhang12** [comment]
Deprecate the old code of `convert_vid_format`,  we could use the neat coco api to realize this function  as the following:
```python
VIS = defaultdict(list)
the 'categories' and 'videos' has already loaded in this dataset, don't need to reload it.
for ins_id, img_ids in self.instancesToImgs.items():
    instance_infos = []
    for img_id in img_ids:
           for ann in self.coco.imgToAnns[img_id]:
                if ann['instance_id'] == instance_id:
                    add some code to load ann information of this instance
    VIS['annotations'].append(instance_infos)
```

File: tests/test_data/test_datasets/test_vis_dataset.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def test_vis_evaluation():
+    dataset_class = DATASETS.get('YouTubeVISDataset')
+    dataset_object = dataset_class(
+        '2019', ann_file=DEMO_ANN_FILE, pipeline=[], test_mode=True)
+    results_json = mmcv.load(DEMO_RES_FILE)
+
+    results = defaultdict(list)
+    np_fmt = []
```
Comment (JingweiZhang12) [inline tests/test_data/test_datasets/test_vis_dataset.py:23]: track_bboxes_numpy = []

File: tests/test_data/test_datasets/test_vis_dataset.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    for i in results_json['track_bboxes']:
+        tmp = []
+        for j in i:
```
Comment (JingweiZhang12) [inline tests/test_data/test_datasets/test_vis_dataset.py:26]: don't use i and j, rename them

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 7]
```diff
@@ -137,3 +145,92 @@ def format_results(self,
...
+        vis_anns['videos'] = copy.deepcopy(self.coco.dataset['videos'])
+
+        len_videos = dict()  # mapping from video_id to video_length
+        for video_id, video_infos in self.coco.vidToImgs.items():
+            len_videos[video_id] = len(video_infos)
+
+        for video_id, ins_ids in self.coco.vidToInstances.items():
+            cur_video_len = len_videos[video_id]
+            for ins_id in ins_ids:
+                segm = [None] * cur_video_len
```
Comment (JingweiZhang12) [inline mmtrack/datasets/youtube_vis_dataset.py:211]: add some description to explain that `segm`、`bbox` and `area` may be None in some frames.

**Review by JingweiZhang12 — APPROVED** [review]

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 8]
```diff
@@ -100,15 +105,15 @@ def format_results(self,
...
-                    collect_data[id]['category_ids'].append(labels[i])
-                    collect_data[id]['scores'].append(bboxes[i][4])
-                    if isinstance(masks[i]['counts'], bytes):
-                        masks[i]['counts'] = masks[i]['counts'].decode()
-                    collect_data[id]['segmentations'][frame_id] = masks[i]
+                    collect_data[id]['category_ids'].append(labels[j])
+                    collect_data[id]['scores'].append(bboxes[j][4])
+                    if isinstance(masks[j]['counts'], bytes):
+                        masks[j]['counts'] = masks[j]['counts'].decode()
+                    collect_data[id]['segmentations'][frame_id] = masks[j]
```
Comment (GT9505) [inline mmtrack/datasets/youtube_vis_dataset.py:116]: pull master to remove the duplicated codes

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 9]
```diff
@@ -53,7 +55,8 @@ def set_dataset_classes(cls, dataset_version):
     def format_results(self,
                        results,
                        resfile_path=None,
-                       metrics=['track_segm']):
+                       metrics=['track_segm'],
+                       return_json=False):
```
Comment (GT9505) [inline mmtrack/datasets/youtube_vis_dataset.py:59]: return_json -> save_as_json
and set the default value to True for save_as_json

File: tests/data/demo_vis_data/results.json [hunk 10]
```diff
@@ -0,0 +1 @@
...00000000000000000000000000000000000000000000000000000000000001O0000000000000000000000000000000000000000000000000000000000000000000000000000000000PVZ10PjeN00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000`ie00`VZO00000000000000000000000000000000000000000000000000000000000000000000000O1000000000000000000000000000000O1000000000000000000000000O100O10000O100O100O1O100O1O1O1O1O1N2N2N2N2O1N2N2O1N2N2O1O1N2O1N200O1O100O1O10000O2O0O10000O10000O100O10000O2O000O10000O10000O10001O0O10000O10000O101O000O101N101N101N1O2O1N1O2N2N2N2M3M4L3M3N2M3M4L3N2M3M3N1N3M3N2N2O0O2N2N2O0O2O1N101N101O1N2O1O1O1N2O1O1N3N1O2M2O2N1O2M3N2N2M3N1O1O1N2O1O1N2O1O001N2O001O0O2O001N101N101N2N101N2N1O2N2N2N2N2M3N3M2N3L4M3M2N2N3M2N2N2N3M2N2N2N1O2N2N2N2O0O2N2O0O2O1N2O1N101N2O1N2O1N2O1O1N2O1N2O1N2O1O1N2O1N2O2M2O1O1N2O1O1N101N2O001N101N101N2O0O2N2O0O2N2N2M3M3M4K4L4L5J5K9A`ib4"}], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []]]}
```
Comment (GT9505) [inline tests/data/demo_vis_data/results.json:1]: same here

File: mmtrack/datasets/youtube_vis_dataset.py [hunk 11]
```diff
@@ -137,3 +145,96 @@ def format_results(self,
...
+        else:
+            raise TypeError('metric must be a list or a str.')
+        allowed_metrics = ['track_segm']
+        for metric in metrics:
+            if metric not in allowed_metrics:
+                raise KeyError(f'metric {metric} is not supported.')
+
+        eval_results = dict()
+        test_results = self.format_results(results, return_json=True)
+        vis_results = self.convert_vis_format()
```
Comment (GT9505) [inline mmtrack/datasets/youtube_vis_dataset.py:175]: `convert_vis_format` -> convert_back_to_vis_format

**Review by GT9505 — APPROVED** [review]

**Pull request merged**